### PR TITLE
rbd: add an initial set of basic mirroring functions

### PR DESCRIPTION
**Pull request by phlogistonjohn**
Add functions:
* SetMirrorMode implementing rbd_mirror_mode_set
* GetMirrorMode implementing rbd_mirror_mode_get
* MirrorEnable implementing rbd_mirror_image_enable2
* MirrorDisable implementing rbd_mirror_image_disable
* MirrorPromote implementing rbd_mirror_image_promote
* MirrorDemote implementing rbd_mirror_image_demote
* MirrorResync implementing rbd_mirror_image_resync
* MirrorInstanceId implementing rbd_mirror_image_get_instance_id

For now, these mirroring related functions are only supported on octopus
builds. Right now the demand is for snapshot based mirroring, which is
new in octopus. We can always relax this in the future and add the
necessary support for nautilus, for journaling only based mirroring.

Part of: #393  - however not all functions are added in this PR.

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code

### CODE REVIEW

File: rbd/mirror.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
+// +build !luminous,!mimic,!nautilus
```
Comment (nixpanic) [inline rbd/mirror.go:1]: we don't have luminous anymore, no need to add it to new files

**Review by nixpanic — CHANGES_REQUESTED** [review]
No need to mention unsupported releases, I think. You might want to change this.
Surprised you included everything in a single commit (OK, but unexpected).

Looks good otherwise!

**Comment by phlogistonjohn** [comment]
> No need to mention unsupported releases, I think. You might want to change this.
> Surprised you included everything in a single commit (OK, but unexpected).
> 
> Looks good otherwise!

FWIW, many of the functions depended on each other for testing, and individually each function's implementation was pretty simple (barring InstanceId but even that's not too bad). The more complex stuff will have to be added (and tested!) in follow up prs.

**Review by agarwal-mudit — DISMISSED** [review]
LGTM

**Comment by phlogistonjohn** [comment]
@Mergifyio rebase

**Comment by phlogistonjohn** [comment]
@mergifyio rebase

File: rbd/mirror.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
+// +build !mimic,!nautilus
```
Comment (nixpanic) [inline rbd/mirror.go:1]: mimic has been dropped in the mean time, not need to mention it in the build-constraints anymore

File: rbd/mirror_test.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	}()
+
+	ioctx, err := conn.OpenIOContext(poolName)
+	assert.NoError(t, err)
+	defer func() {
+		ioctx.Destroy()
+	}()
+
+	// verify that mirroring is not enabled on this new pool
+	m, err := GetMirrorMode(ioctx)
```
Comment (nixpanic) [inline rbd/mirror_test.go:34]: can you add a call to `GetMirrorMode()` with an invalid/nil context too?

**Review by nixpanic — CHANGES_REQUESTED** [review]
I'd like to see `GetMirrorMode()` tested a little more, if reasonable.

The mimic removal from the build-constraints is not required (but would be nice), that is something we should probably address in existing files anyway.

File: rbd/mirror.go (hunk 2 above)
Comment (phlogistonjohn) [reply rbd/mirror.go]: done

File: rbd/mirror_test.go (hunk 3 above)
Comment (phlogistonjohn) [reply rbd/mirror_test.go:70]: done, with caveats (please see #427)

**Comment by phlogistonjohn** [comment]
@Mergifyio rebase

File: rbd/mirror.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	return getError(ret)
+}
+
+// MirrorInstanceId returns a string naming the instance id for the image.
+//
+// Implements:
+//  int rbd_mirror_image_get_instance_id(rbd_image_t image,
+//                                       char *instance_id,
+//                                       size_t *id_max_length);
+func (image *Image) MirrorInstanceId() (string, error) {
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror.go:144]: I don't care for go-lint because it is too rigid IMO. The codebase already used Id in many places when we started maintaingin go-ceph. But unfortunately we're not very consistent here:
```
$ grep -IR 'func ' rbd | grep -c  ID
9
$ grep -IR 'func ' rbd | grep -c  Id
5
```
But since we're already inconsistent here and ID has the edge over Id and the tool complains I won't fight it. :-)

File: rbd/mirror_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
+// +build !mimic,!nautilus
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror_test.go:1]: I can drop it because we no longer support mimic (or luminous) and it being there is largely an artifact of when I started working on this.

File: rbd/mirror.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+// SetMirrorMode is used to enable or disable pool level mirroring with either
+// an automatic or per-image behavior.
+//
+// Implements:
+//  int rbd_mirror_mode_set(rados_ioctx_t io_ctx,
+//                          rbd_mirror_mode_t mirror_mode);
+func SetMirrorMode(ioctx *rados.IOContext, mode PoolMirrorMode) error {
+	ret := C.rbd_mirror_mode_set(
+		cephIoctx(ioctx),
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror.go:54]: Yeah, I can add a test for ioctx == nil -> panic. We can also add separate tests (outside this pr I'd suggest) to test the behavior of cephIoctx and ioctx.Pointer() for these conditions.

File: rbd/mirror.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import "C"
+
+import (
+	"unsafe"
+
+	"github.com/ceph/go-ceph/internal/retry"
+	"github.com/ceph/go-ceph/rados"
+)
+
+// PoolMirrorMode is used to indicate the mirroring approach for a pool.
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror.go:22]: Sorry but I'm having difficulty parsing, "I would drop the "Pool" in analogy to the defined constants." Do you suggest that the prefix "Pool" in the type name is misleading and to rename "PoolMirrorMode" -> "MirrorMode"?
If so, I have no issues with that. If not please let me know.

File: rbd/mirror.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror.go]: changed

File: rbd/mirror.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply rbd/mirror.go]: assuming my previous comment was correct. I made the mentioned changes. Let me know if I misinterpreted.

[the remark this answers is not in the document]
Comment (ansiwen) [reply rbd/mirror.go]: Sorry, yes, your interpretation was 100% spot-on. :-)

**Review by ansiwen — DISMISSED** [review]
LGTM

**Comment by phlogistonjohn** [comment]
@agarwal-mudit or @nixpanic PR was updated so this needs re-review. Thanks.

**Comment by phlogistonjohn** [comment]
@mergifyio rebase

**Comment by phlogistonjohn** [comment]
@ansiwen can you please reapprove? You left a LGTM and I think a gh approval but mergify has dissmissed it. I don't think I made any new changes since, but feel free to give it a 2nd look even though I think mergify is in the wrong.

**Review by ansiwen — APPROVED** [review]

File: rbd/mirror.go [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+)
+
+// ImageMirrorMode is used to indicate the mirroring approach for an RBD image.
+type ImageMirrorMode int64
+
+const (
+	// ImageMirrorModeJournal uses journaling to propagate RBD images between
+	// ceph clusters.
+	ImageMirrorModeJournal = ImageMirrorMode(C.RBD_MIRROR_IMAGE_MODE_JOURNAL)
+	// ImageMirrorModeSnapshot uses snapshot RDB images to propagate images
```
Comment (agarwal-mudit) [inline rbd/mirror.go:41]: Nit: 
```suggestion
	// ImageMirrorModeSnapshot uses snapshot to propagate images

**Review by agarwal-mudit — APPROVED** [review]

File: rbd/mirror.go (hunk 8 above)
Comment (phlogistonjohn) [reply rbd/mirror.go:41]: I'll fix that in a follow up PR. I want to get this merged now that I have the needed # of reviews and dont want mergify mucking with this again.
